### PR TITLE
Pick tolerations applied to cephtoolbox on ocsinitialization & structure the cephtoolbox tolerations on storagecluster CR

### PR DESCRIPTION
**Pull request by malayparida2000**
Resolves-https://bugzilla.redhat.com/show_bug.cgi?id=2131220

In 4.11 in this [#1602](https://github.com/red-hat-storage/ocs-operator/pull/1602), We moved the toolbox enable/disable & reconcile logic to the storagecluster side. But the implementation to still pick up tolerations added to the ocsinitialization CR was not there. So adding it now.

The tolerations on the same thing on the storagecluster side are also being structured properly to maintain uniformity.
And also adding missing implementation for the reconcile strategy.

### CODE REVIEW

File: api/v1/storagecluster_types.go [hunk 1]
```diff
@@ -211,13 +211,10 @@ type ManageCephObjectStoreUsers struct {
...
+// ManageCephToolbox defines how to reconcile Ceph toolbox & the placement configuration
 type ManageCephToolbox struct {
-	ReconcileStrategy string `json:"reconcileStrategy,omitempty"` // default behavior here is effectively `ignore`
-
-	// Tolerations if specified set toolbox ceph tools pod tolerations
-	// Defaults to empty
-	// +optional
-	Tolerations []corev1.Toleration `json:"cephTolerations,omitempty"`
+	ReconcileStrategy string               `json:"reconcileStrategy,omitempty"`
+	Placement         rookCephv1.Placement `json:"placement,omitempty"`
```
Comment (umangachapagain) [inline api/v1/storagecluster_types.go:217]: We need to move this to `StorageClusterSpec.Placement` if we want to be consistent with other components.
None of the Manage* struct have a placement spec inside it.

File: controllers/storagecluster/cephtoolbox.go [hunk 2]
```diff
@@ -31,11 +31,21 @@ func (r *StorageClusterReconciler) ensureToolsDeployment(sc *ocsv1.StorageCluste
...
+	ocsinit := &ocsv1.OCSInitialization{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: "ocsinit", Namespace: namespace}, ocsinit)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	// Get tolerations from both the storagecluster CR & ocsinitialization CR
+	// Keeping this for ability to add tolerations from ocsinitialization CR
+	tolerations = append(tolerations, sc.Spec.ManagedResources.CephToolbox.Placement.Tolerations...)
+	tolerations = append(tolerations, ocsinit.Spec.Tolerations...)
```
Comment (umangachapagain) [inline controllers/storagecluster/cephtoolbox.go:50]: It'd be better to merge it such that there are no duplicate toleration.

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (malayparida2000) [reply api/v1/storagecluster_types.go:217]: So we can just take the Placement info from the Storagecluster Spec.Placement.
We can take the tolerations out of that & apply that onto the toolbox. That would work right?
We would no longer need this field for Placement/Toleration inside the ManageCephToolBox.

Comment (malayparida2000) [reply api/v1/storagecluster_types.go]: Done

File: controllers/storagecluster/cephtoolbox.go (hunk 2 above)
Comment (malayparida2000) [reply controllers/storagecluster/cephtoolbox.go]: Done

File: api/v1/storagecluster_types.go [hunk 3]
```diff
@@ -211,13 +211,9 @@ type ManageCephObjectStoreUsers struct {
 	ReconcileStrategy string `json:"reconcileStrategy,omitempty"`
 }
 
+// ManageCephToolbox defines how to reconcile Ceph toolbox & the placement configuration
```
Comment (umangachapagain) [inline api/v1/storagecluster_types.go:214]:
```suggestion
// ManageCephToolbox defines how to reconcile Ceph toolbox
```

File: controllers/storagecluster/cephtoolbox.go [hunk 4]
```diff
@@ -76,3 +87,15 @@ func (r *StorageClusterReconciler) ensureToolsDeployment(sc *ocsv1.StorageCluste
 	}
 	return nil
 }
+
+func removeDuplicateTolerations(tolerations []corev1.Toleration) []corev1.Toleration {
```
Comment (umangachapagain) [inline controllers/storagecluster/cephtoolbox.go:91]: please add unit test for this

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: api/v1/storagecluster_types.go (hunk 3 above)
Comment (malayparida2000) [reply api/v1/storagecluster_types.go]: Done

**Comment by malayparida2000** [comment]
@umangachapagain PTAL, We need to merge this today as this is targeted for 4.11.3

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**